backend/fs/EntityWithSourcesFS.cpp

```diff
@@ -97,38 +97,35 @@ std::shared_ptr<base::ISource> EntityWithSourcesFS::getSource(const size_t index
 }
 
 void EntityWithSourcesFS::sources(const std::vector<Source> &sources) {
-    // extract vectors of ids from vectors of new & old sources
-    std::vector<std::string> ids_new(sources.size());
-    std::transform(sources.begin(), sources.end(), ids_new.begin(), util::toId<Source>);
-    // FIXME: issue #473
-    // FIXME: this could go to a front-end like thing
-    std::vector<Source> sources_old(static_cast<size_t>(sourceCount()));
-    for (size_t i = 0; i < sources_old.size(); i++) sources_old[i] = getSource(i);
-    std::vector<std::string> ids_old(sources_old.size());
-    std::transform(sources_old.begin(), sources_old.end(), ids_old.begin(), util::toId<Source>);
-    // sort them
-    std::sort(ids_new.begin(), ids_new.end());
-    std::sort(ids_old.begin(), ids_old.end());
-    // get ids only in ids_new (add), ids only in ids_old (remove) & ignore rest
-    std::vector<std::string> ids_add;
-    std::vector<std::string> ids_rem;
-    std::set_difference(ids_new.begin(), ids_new.end(), ids_old.begin(), ids_old.end(),
-                        std::inserter(ids_add, ids_add.begin()));
-    std::set_difference(ids_old.begin(), ids_old.end(), ids_new.begin(), ids_new.end(),
-                        std::inserter(ids_rem, ids_rem.begin()));
-
-    // check if all new sources exist
-    Block tmp(entity_block);
-    auto found = tmp.findSources(util::IdsFilter<Source>(ids_add));
-    if (ids_add.size() != found.size())
-        throw std::runtime_error("One or more sources do not exist in this block!");
-    // add sources
-    for (auto id : ids_add) {
-        addSource(id);
+    auto cmp = [](const Source &a, const Source& b) { return a.name() < b.name(); };
+
+    std::vector<Source> new_arrays(sources);
+
+    ndsize_t current = sourceCount();
+    size_t count = nix::check::fits_in_size_t(current, "entityCount() failed; count > size_t.");
+    std::vector<Source> old_arrays(count);
+
+    //check if this can be replaced
+    for (size_t i = 0; i < old_arrays.size(); i++) {
+        old_arrays[i] = getSource(i);
+    }
+
+    std::sort(new_arrays.begin(), new_arrays.end(), cmp);
+    std::sort(old_arrays.begin(), old_arrays.end(), cmp);
+    std::vector<Source> add;
+    std::vector<Source> rem;
+
+    std::set_difference(new_arrays.begin(), new_arrays.end(), old_arrays.begin(),
+                        old_arrays.end(), std::inserter(add, add.begin()), cmp);
+    std::set_difference(old_arrays.begin(), old_arrays.end(), new_arrays.begin(),
+                        new_arrays.end(), std::inserter(rem, rem.begin()), cmp);
+
+    for (const auto &e : add) {
+        addSource(e.id());
```

Comment:
Is there are a reason we can't just pass the whole entity (i.e. `e`); same below.